orbital-camera.ts

```diff
@@ -0,0 +1,230 @@
+        this.object.getPositionWorld(this.origin);
+        this.updateCamera();
+    }
+
+    onActivate(): void {
+        const canvas = this.engine.canvas;
+
+        canvas.addEventListener('mousemove', this.onMouseMove);
+        if (this.mouseButtonIndex === 2) {
+            canvas.addEventListener('contextmenu', preventDefault, false);
+        }
+        canvas.addEventListener('mousedown', this.onMouseDown);
+        canvas.addEventListener('mouseup', this.onMouseUp);
+        canvas.addEventListener('wheel', this.onMouseScroll);
+
+        canvas.addEventListener('touchstart', this.onTouchStart, {passive: false});
+        canvas.addEventListener('touchmove', this.onTouchMove, {passive: false});
+        canvas.addEventListener('touchend', this.onTouchEnd);
+    }
+
+    onDeactivate(): void {
+        const canvas = this.engine.canvas;
+
+        canvas.removeEventListener('mousemove', this.onMouseMove);
+        if (this.mouseButtonIndex === 2) {
+            canvas.removeEventListener('contextmenu', preventDefault, false);
+        }
+        canvas.removeEventListener('mousedown', this.onMouseDown);
+        canvas.removeEventListener('mouseup', this.onMouseUp);
+        canvas.removeEventListener('wheel', this.onMouseScroll);
+
+        canvas.removeEventListener('touchstart', this.onTouchStart);
+        canvas.removeEventListener('touchmove', this.onTouchMove);
+        canvas.removeEventListener('touchend', this.onTouchEnd);
+
+        // Reset state to make sure nothing gets stuck
+        this.mouseDown = false;
+        this.initialPinchDistance = 0;
+        this.touchStartX = 0;
+        this.touchStartY = 0;
+
+        this.azimuthSpeed = 0;
+        this.polarSpeed = 0;
+    }
+
+    update(): void {
+        if (!this.mouseDown) {
+            // Apply deceleration only when the user is not actively dragging
+            this.azimuthSpeed *= this.decelerationFactor;
+            this.polarSpeed *= this.decelerationFactor;
+
+            // Stop completely if the speed is very low to avoid endless tiny movements
+            if (Math.abs(this.azimuthSpeed) < 0.01) this.azimuthSpeed = 0;
+            if (Math.abs(this.polarSpeed) < 0.01) this.polarSpeed = 0;
+        }
+
+        // Apply the speed to the camera angles
+        this.azimuth += this.azimuthSpeed;
+        this.polar += this.polarSpeed;
+
+        // Clamp the polar angle
+        this.polar = Math.min(this.maxElevation, Math.max(this.minElevation, this.polar));
+
+        // Update the camera if there's any speed
+        if (this.azimuthSpeed !== 0 || this.polarSpeed !== 0) {
+            this.updateCamera();
+        }
+    }
+
+    private updateCamera() {
+        this.object.setPositionWorld([
+            this.radial * Math.sin(deg2rad(this.azimuth)) * Math.cos(deg2rad(this.polar)),
+            this.radial * Math.sin(deg2rad(this.polar)),
+            this.radial * Math.cos(deg2rad(this.azimuth)) * Math.cos(deg2rad(this.polar)),
+        ]);
```

Comment:
This is creating a temporary minimal garbage list with the `[]` literal.

You could move that into a temp vector, and fill it before passing to setPositionWorld:

```js
tempVec[0] = ...;
tempVec[1] = ...;
tempVec[2] = ...;
setPositionWorld(tempVec);
```

Reply:
Fixed.